File: sort/java/QuickSort.java
public class QuickSort {
    public static void sort(int[] array) {
        sort(array, 0, array.length - 1);
    }

    private static void sort(int[] array, int low, int high) {
        if (low >= high) {
            return;
        }

        int pivot = array[high];
        int pos = low;

        for (int i = low; i < high; i++) {
            if (array[i] < pivot) {
                Utils.swap(array, pos, i);
                pos++;
            }
        }

        Utils.swap(array, pos, high);

        sort(array, low, pos - 1);
        sort(array, pos + 1, high);
    }
}

File: sort/c/quicksort.c
#include "algorithms.h"

void quicksort(int *array, size_t low, size_t high) {
  if (low >= high) {
    return;
  }

  int pivot = array[high];
  size_t pos = low;

  for (size_t i = low; i < high; i++) {
    if (array[i] < pivot) {
      swap(&array[pos], &array[i]);
      pos++;
    }
  }

  swap(&array[pos], &array[high]);

  if (pos > 0) {
    quicksort(array, low, pos - 1);
  }
  quicksort(array, pos + 1, high);
}
